src/read/util.rs

```diff
@@ -0,0 +1,174 @@
+use core::fmt;
+use core::mem::MaybeUninit;
+use core::ops;
+use core::ptr;
+use core::slice;
+
+// Use a helper trait since const generics can't be used due to MSRV.
+// Implementer must ensure that `Storage` is `[MaybeUninit<Item>; CAPACITY]`
+pub unsafe trait Array {
+    type Item;
+    type Storage;
+
+    const CAPACITY: usize;
+
+    fn as_slice(storage: &Self::Storage) -> &[MaybeUninit<Self::Item>];
+    fn as_mut_slice(storage: &mut Self::Storage) -> &mut [MaybeUninit<Self::Item>];
+}
+
+macro_rules! impl_array {
+    () => {};
+    ($n:literal $($rest:tt)*) => {
+        // SAFETY: `Storage` is `[MaybeUninit<Item>; CAPACITY]`.
+        unsafe impl<T> Array for [T; $n] {
+            type Item = T;
+            type Storage = [MaybeUninit<T>; $n];
+            const CAPACITY: usize = $n;
+
+            fn as_slice(storage: &Self::Storage) -> &[MaybeUninit<T>] {
+                storage
+            }
+
+            fn as_mut_slice(storage: &mut Self::Storage) -> &mut [MaybeUninit<T>] {
+                storage
+            }
+        }
+        impl_array!($($rest)*);
+    }
+}
+
+impl_array!(4 192);
+
+pub struct ArrayVec<A: Array> {
```

Comment:
Use `pub(crate)` for top-level items to make it clearer that we never re-export anything here.